README.md

```diff
@@ -1,2 +1,1088 @@
+
+The entrypoint's value should be a class inheriting from the
+`sim.front.Renderer` class.
+
+The `default` renderer uses this method to register itself as well.
+
+For example, let's add a "better" renderer for the ***Macro**mouse*
+competition.
+
+First, let's implement it in our package, `macromouse.py`:
+
+```python
+from sim.gui import GUIRenderer
+
+# GUIRenderer inherits from Renderer so we don't have to do so explicitly.
+class MacroRenderer(GUIRenderer):
+    """The best renderer."""
+
+    def __init__(self, sim: Simulator):
+        super().__init__(sim)
+        pg.display.set_caption('Macromouse')
+```
+
+Now, we'll add it in our `pyproject.toml`:
+
+```toml
+# ...
+[project.entry-points."micromouse.gui"]
+better = "macromouse:MacroRenderer"
+# ...
+```
+
+After installing our new package, we can use the new engine:
+
+```sh
+$ usage: micromouse sim [-h] [-m MAZE] [-s START_POS] [-d START_DIRECTION] [-g GOALS] [-p PRESET] [-e {default,better}]
+
+Run the simulator.
+
+options:
+  -h, --help            show this help message and exit
+  -m MAZE, --maze MAZE  The maze to load. May be a file or a maze input. (default: 1x1 maze)
+  -s START_POS, --start-pos START_POS
+                        The starting position. (default: the top-left corner - (0, 0))
+  -d START_DIRECTION, --start-direction START_DIRECTION
+                        The starting direction. (default: a valid direction for the starting position)
+  -g GOALS, --goals GOALS
+                        The goal positions. (default: the bottom-right corner - (M-1, N-1) in a MxN maze)
+  -p PRESET, --preset PRESET
+                        A maze+start+goals preset to load.
+  -e {default,better}, --engine {default,better}
+                        The render engine to use. (default: a simple pygame GUI renderer, requires the 'gui' feature)
+```
+
+(note that the new engine appeared as an option)
+
+## Tools
+
+This package also allows bundling various Micromouse-related tools.
+
+Tools are accessible via the `micromouse tool` command as subcommands:
+
+```sh
+$ micromouse tool -h
+usage: micromouse tool [-h] {maze} ...
+
+Run a tool.
+
+options:
+  -h, --help  show this help message and exit
+
+tools:
+  All registered tools.
+
+  {maze}
+    maze      Utilities for viewing and manipulating a maze.
+```
+
+### Provided Tool(s)
+
+#### Maze Editor
+
+This package provides a *MazeEditor* tool for basic creation and manipulation
+of mazes.
+
+```sh
+$ micromouse tool maze -h
+usage: micromouse tool maze [-h] {generate,render,rotate,transpose,flip} ...
+
+Utilities for viewing and manipulating a maze.
+
+options:
+  -h, --help            show this help message and exit
+
+subcommands:
+  {generate,render,rotate,transpose,flip}
+    generate            Generate a new maze.
+    render              Render a maze (can be used to convert other formats into .maze format).
+    rotate              Rotate a maze (can be used to convert other formats into .maze format).
```

Comment:
Not sure how rotate converts other formats into .maze format and if so why don't the other functions can do the same? Is there something special in rotate?

Reply:
Rotate 0 turns. The others also "convert" but are incapable of no-op.